src/traces/table/calc.js

```diff
@@ -0,0 +1,131 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var c = require('./constants');
+var wrap = require('../../lib/gup').wrap;
+var extendFlat = require('../../lib/extend').extendFlat;
+
+module.exports = function calc(gd, trace) {
+    var domain = trace.domain;
+    var groupWidth = Math.floor(gd._fullLayout._size.w * (domain.x[1] - domain.x[0]));
+    var groupHeight = Math.floor(gd._fullLayout._size.h * (domain.y[1] - domain.y[0]));
+    var headerRowHeights = trace.header.values[0].map(function() {return trace.header.height;});
+    var rowHeights = trace.cells.values[0].map(function() {return trace.cells.height;});
+    var headerHeight = headerRowHeights.reduce(function(a, b) {return a + b;}, 0);
+    var scrollHeight = groupHeight - headerHeight;
+    var minimumFillHeight = scrollHeight + c.uplift;
+    var anchorToRowBlock = makeAnchorToRowBlock(rowHeights, minimumFillHeight);
+    var anchorToHeaderRowBlock = makeAnchorToRowBlock(headerRowHeights, headerHeight);
+    var headerRowBlocks = makeRowBlock(anchorToHeaderRowBlock, []);
+    var rowBlocks = makeRowBlock(anchorToRowBlock, headerRowBlocks);
+    var uniqueKeys = {};
+    var columnOrder = trace._fullInput.columnorder;
+    var columnWidths = trace.header.values.map(function(d, i) {
+        return Array.isArray(trace.columnwidth) ?
+            trace.columnwidth[Math.min(i, trace.columnwidth.length - 1)] :
+            isFinite(trace.columnwidth) && trace.columnwidth !== null ? trace.columnwidth : 1;
+    });
+    var totalColumnWidths = columnWidths.reduce(function(p, n) {return p + n;}, 0);
+
+    // fit columns in the available vertical space as there's no vertical scrolling now
+    columnWidths = columnWidths.map(function(d) {return d / totalColumnWidths * groupWidth;});
+
+    var calcdata = {
+        key: trace.index,
+        translateX: domain.x[0] * gd._fullLayout._size.w,
+        translateY: gd._fullLayout._size.h - domain.y[1] * gd._fullLayout._size.h,
+        size: gd._fullLayout._size,
+        width: groupWidth,
+        height: groupHeight,
+        columnOrder: columnOrder, // will be mutated on column move, todo use in callback
```

Comment:
Still TODO?

Reply:
I thought that a callback on column rearrangement (and scroll) would be useful but as I didn't get this as a piece of requirement, more pressing things took my time. I left in comments to revisit, whether in this PR or in a follow-up PR. Due to the target date, a follow-up PR would be safer.

Reply:
Thanks for the info. No action required. :+1: